### ISSUE TITLE
Failed to collect raid controller device S.M.A.R.T data

### ISSUE THREAD

**Issue by sdragon83** [prometheus-community/smartctl_exporter#89]
I tried to collect data from a server with a raid controller through smartctl exporter.

However, an error occurred as below.

How can i collect S.M.A.R.T data on raid controller devices?

![image](https://user-images.githubusercontent.com/17839907/196605969-3f5341c0-c7e7-4a48-9857-119cca521262.png)

**Comment by tomazb** [comment]
Yes, this is the real reason why you need such a service in the first place - to monitor devices that are not easily visible inside the operating system.

**Comment by josefzahner** [comment]
@marpears I can read the device info with smartctl including the device option, but NOT with smartctl_exporter...

`$smartctl --json --info --health --attributes --tolerance=verypermissive --nocheck=standby --format=brief --device cciss,1 /dev/sdb
{
  "json_format_version": [
    1,
    0
  ],
  "smartctl": {
    "version": [
      7,
      3
    ],
    "svn_revision": "5338",
    "platform_info": "x86_64-linux-3.10.0-957.27.2.el7.x86_64",
    "build_info": "(local build)",
    "argv": [
      "smartctl",
      "--json",
...
`

but this doesn't work:
`$ smartctl_exporter --smartctl.device='cciss,1 /dev/sdb'
ts=2022-12-02T09:40:45.718Z caller=main.go:90 level=info msg="Starting smartctl_exporter" version="(version=0.9.1, branch=HEAD, revision=a58c632ea8fa0f4f10a9ac9e941e610a7bb2efc1)"
ts=2022-12-02T09:40:45.718Z caller=main.go:91 level=info msg="Build context" build_context="(go=go1.19.3, user=root@fa2a9a938fb5, date=20221106-21:46:18)"
ts=2022-12-02T09:40:45.735Z caller=main.go:112 level=warn msg="Device unavailable" name="cciss,1 /dev/sdb"
ts=2022-12-02T09:40:45.735Z caller=main.go:119 level=info msg="No devices specified, trying to load them automatically"
ts=2022-12-02T09:40:45.735Z caller=main.go:124 level=error msg="No devices found"
`

**Comment by kfox1111** [comment]
how does one configure cciss,1? I need to do it on some of my nodes and have not found a way yet.

**Comment by anthonyeleven** [comment]
This is a gating factor for me too.  I've added comments to the above issue and linked PR.

**Comment by jakubgs** [comment]
This is also an issue for me. I guess a proper solution would involve adding a separate flag to provide extra flags for `smartctl`.

**Comment by anthonyeleven** [comment]
The tool should discover such HBAs and do so automagically at per-device granularity, since there can and will be a mixed population of direct-attach, passthrough, and hidden-by-VD drives on various sytems and especially within a given system.

`smartmon.sh` for example does this:

```device_list="$($SMARTCTL --scan-open | awk '/^\/dev/{print $1 "|" $3}')"

for device in ${device_list}; do
  disk="$(echo ${device} | cut -f1 -d'|')"
  type="$(echo ${device} | cut -f2 -d'|')"
  active=1
  echo "smartctl_run{disk=\"${disk}\",type=\"${type}\"}" "$(TZ=UTC date '+%s')"
  # Check if the device is in a low-power mode
  $SMARTCTL -n standby -d "${type}" "${disk}" > /dev/null || active=0
  echo "device_active{disk=\"${disk}\",type=\"${type}\"}" "${active}"
  # Skip further metrics to prevent the disk from spinning up
  test ${active} -eq 0 && continue
  # Get the SMART information and health
  $SMARTCTL  -i -H -d "${type}" "${disk}" | parse_smartctl_info "${disk}" "${type}"
  # Get the SMART attributes
  case ${type} in
  sat) $SMARTCTL -A -d "${type}" "${disk}" | parse_smartctl_attributes "${disk}" "${type}" ;;
  sat+megaraid*) $SMARTCTL -A -d "${type}" "${disk}" | parse_smartctl_attributes "${disk}" "${type}" ;;
  scsi) $SMARTCTL -A -d "${type}" "${disk}" | parse_smartctl_scsi_attributes "${disk}" "${type}" ;;
  nvme) $SMARTCTL -A -d "${type}" "${disk}" | parse_smartctl_nvme_attributes "${disk}" "${type}" ;;
  megaraid*) $SMARTCTL -A -d "${type}" "${disk}" | parse_smartctl_scsi_attributes "${disk}" "${type}" ;;
  *)
    echo "disk type is not sat, scsi or megaraid but ${type}"
    exit
    ;;
  esac
done | format_output```


Mind you, I *despise* RoC HBAs and would just as soon never have one, or to set passthrough/JBOD on legacy systems, but walking into an existing deployment of thousands I don't have the luxury of greenfield.

**Comment by anthonyeleven** [comment]
@jakubgs It's more than just extra flags, it's discovery too.
```[root@bridge-2 anthony.datri]# /opt/index/sbin/smartctl --scan
/dev/sda -d scsi # /dev/sda, SCSI device
/dev/bus/0 -d megaraid,0 # /dev/bus/0 [megaraid_disk_00], SCSI device
/dev/bus/0 -d megaraid,1 # /dev/bus/0 [megaraid_disk_01], SCSI device```

I no longer have HP HBAs, but it would be polite for however this is done to be architected in such a way that they could be supported later.

I hope to sunset RoC VDs through attrition, but that will take years :-/

**Comment by kfox1111** [comment]
Any way to do this yet?

**Comment by anthonyeleven** [comment]
I’d do it myself if I had the coding skills.  It really is a fatal flaw.  Mind you HBA RAID is itself a fatal flaw but Dell’s BOSS-N1 is too useful, though one has to invoke ‘mvcli’ to get status.  On Jul 14, 2023, at 5:10 PM, kfox1111 ***@***.***> wrote:﻿
Any way to do this yet?

**Comment by jakubgs** [comment]
Main issue as far as I can tell is that even if you discover the devices, often you won't get much info from them:
```json
{
  "json_format_version": [1, 0],
  "smartctl": {
    "version": [7, 2],
    "svn_revision": "5155",
    "platform_info": "x86_64-linux-5.15.0-79-generic",
    "build_info": "(local build)",
    "argv": ["smartctl", "-A", "--device", "cciss,1", "/dev/sdb", "--json"],
    "exit_status": 0
  },
  "device": {
    "name": "/dev/sdb",
    "info_name": "/dev/sdb [cciss_disk_01] [SCSI]",
    "type": "cciss",
    "protocol": "SCSI"
  },
  "temperature": {
    "current": 21,
    "drive_trip": 70
  },
  "power_on_time": {
    "hours": 47138,
    "minutes": 5
  },
  "scsi_grown_defect_list": 0
}
```
Temperatur and power on time... not great.

**Comment by anthonyeleven** [comment]
Better than nothing, but yeah.  I haven't had an HP HBA to work with for years, but re the `scsi` factor above, is the subject drive SAS?  I would not be surprised if this would not surface SATA (but it might).

**Comment by anthonyeleven** [comment]
I'm increasingly leaning toward having a protege write a SMART harvest from scratch in Python, which would make it easier to normalize the vagaries of data that `smartctl` gives us.

**Comment by jakubgs** [comment]
Personally I'd rather fix what we have working than try from scratch. I'm busy enough dealing with what I have working already have the time to reinvent wheels. Even if I get just temp and power-on hours that's better than deployed SMART exporter just failing at startup and Prometheus returning alerts for the downed service.

**Comment by jakubgs** [comment]
Bu your point about SATA/SAS is well made. I will have to check how that is done on my servers.
